docs/builder/ref.md

```diff
@@ -0,0 +1,290 @@
+# Builder Library Reference
+
+## Top-Level Program Structure
+
+Here's the general structure of a program that uses the builder to generate Calyx code.
+
+```python
+# import the builder library
+import calyx.builder as cb
+
+
+# define my_subcomponent
+def add_my_subcomponent(prog):
+    # subcomponent definition here
+
+
+# define my_component
+def my_component(prog, my_subcomponent): 
+    # add the component to the program
+    my_component = prog.component("my_component")
+
+    # add my_subcomponent as a cell of my_component
+    my_subcomponent = my_component.cell("my_subcomponent", my_subcomponent)
+
+    # define a my_component group
+    with my_component.group("my_group") as my_group:
+      # group assignments here 
+
+  my_component.control += [my_group]
+
+
+# assemble the program
+def build():
+    prog = cb.Builder()
+    my_subcomponent = add_my_subcomponent(prog)
+    add_main(prog, my_subcomponent)
+
+    # return the generated program
+    return prog.program
+
+
+# emit the program
+if __name__ == "__main__":
+    build().emit()
+```
+
+## Components
+
+### Defining Components
+
+You can define a component by calling the `Builder().component()` method.
+
+```python
+prog = cb.Builder()
+prog.component("my_component")
+```
+
+Components can also be instantiated with a list of cells.
+
+```python
+
+ adders = [adder_1, adder_2, adder_3, adder_4] = [
+        my_component.add(name, 32) for name in ["1", "2", "3", "4"]
+    ]
+my_component = prog.component("my_component", cells=adders)
```

Comment:
ahh gotcha, that makes sense. we probably don't need to include the `cells=` construct for now since it seems that folks at the tutorial are unlikely to use it. thank you!

Reply:
Agreed, we can briefly mention that it exists, but I don't think it's necessary to talk about